dot/parachain/prospective-parachains/fragment-chain/fragment_chain.go

```diff
@@ -0,0 +1,1195 @@
+			}
+		}
+	}
+
+	// to make sure that cycles dont make us loop forever, keep track
+	// of the visited parent head hashes
+	visited := map[common.Hash]struct{}{}
+
+	for queue.Len() > 0 {
+		// queue.PopFront()
+		parent := queue.Remove(queue.Front()).(queueItem)
+		visited[parent.hash] = struct{}{}
+
+		children, ok := storage.byParentHead[parent.hash]
+		if !ok {
+			continue
+		}
+
+		// cannot remove while iterating so store them here temporarily
+		var toRemove []parachaintypes.CandidateHash
+
+		for childHash := range children {
+			child, ok := storage.byCandidateHash[childHash]
+			if !ok {
+				continue
+			}
+
+			// already visited this child. either is a cycle or multipath that lead
+			// to the same candidate. either way, stop this branch to avoid looping
+			// forever
+			if _, ok = visited[child.outputHeadDataHash]; ok {
+				continue
+			}
+
+			// only keep a candidate if its full ancestry was already kept as potential
+			// and this candidate itself has potential
+			if parent.hasPotential && f.checkPotential(child) == nil {
+				queue.PushBack(queueItem{hash: child.outputHeadDataHash, hasPotential: true})
+			} else {
+				// otherwise, remove this candidate and continue looping for its children
+				// but mark the parent's potential as false. we only want to remove its children.
+				toRemove = append(toRemove, childHash)
+				queue.PushBack(queueItem{hash: child.outputHeadDataHash, hasPotential: false})
+			}
+		}
+
+		for _, hash := range toRemove {
+			storage.removeCandidate(hash)
+		}
+	}
+}
+
+type possibleChild struct {
+	fragment           *inclusionemulator.Fragment
+	candidateHash      parachaintypes.CandidateHash
+	outputHeadDataHash common.Hash
+	parentHeadDataHash common.Hash
+}
+
+// populateChain populates the fragment chain with candidates from the supplied `CandidateStorage`.
+// Can be called by the `NewFragmentChain` or when backing a new candidate. When this is called
+// it may cause the previous chain to be completely erased or it may add more than one candidate
+func (f *FragmentChain) populateChain(storage *CandidateStorage) {
+	var cumulativeModifications *inclusionemulator.ConstraintModifications
+	if len(f.bestChain.chain) > 0 {
+		lastCandidate := f.bestChain.chain[len(f.bestChain.chain)-1]
+		cumulativeModifications = lastCandidate.cumulativeModifications.Clone()
+	} else {
+		cumulativeModifications = inclusionemulator.NewConstraintModificationsIdentity()
+	}
+
+	earliestRelayParent := f.earliestRelayParent()
+	if earliestRelayParent == nil {
+		return
+	}
+
+	for len(f.bestChain.chain) < int(f.scope.maxDepth) {
+		childConstraints, err := f.scope.baseConstraints.ApplyModifications(cumulativeModifications)
+		if err != nil {
+			// TODO: include logger
+			fmt.Println("failed to apply modifications:", err)
+			break
+		}
+
+		requiredHeadHash, err := childConstraints.RequiredParent.Hash()
+		if err != nil {
+			panic(fmt.Sprintf("failed while hashing required parent: %s", err.Error()))
```

Comment:
Is this supposed to be a panic?

Reply:
yeah, if the we cannot get the hash of the required parent in a given scope constraint than the whole state is broken